core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -114,119 +117,138 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
     {
-        state.addLong(1); // row counter
-
-        state.setNotNull();
+        long[] decimal = decimalState.getArray();
+        int decimalOffset = decimalState.getArrayOffset();
 
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        counterState.setValue(counterState.getValue() + 1);
+        decimalState.setNotNull();
 
         long rightHigh = block.getLong(position, 0);
         long rightLow = block.getLong(position, SIZE_OF_LONG);
 
         long overflow = addWithOverflow(
-                decimal[offset],
-                decimal[offset + 1],
+                decimal[decimalOffset],
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        if (overflow != 0) {
+            overflowState.setNull(false);
+            overflowState.setValue(overflowState.getValue() + overflow);
+        }
     }
 
-    public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
+    public static void combine(Int128State decimalState, LongState counterState, NullableLongState overflowState, Int128State otherDecimalState, LongState otherCounterState, NullableLongState otherOverflowState)
     {
-        state.addLong(otherState.getLong()); // row counter
-
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        // TODO: kiedy tak jest?
+        if (!decimalState.isNotNull() && !otherDecimalState.isNotNull()) {
+            return;
+        }
 
-        long[] otherDecimal = otherState.getDecimalArray();
-        int otherOffset = otherState.getDecimalArrayOffset();
+        long[] decimal = decimalState.getArray();
+        int decimalOffset = decimalState.getArrayOffset();
+        long[] otherDecimal = otherDecimalState.getArray();
+        int otherDecimalOffset = otherDecimalState.getArrayOffset();
 
-        if (state.isNotNull()) {
+        if (decimalState.isNotNull()) {
             long overflow = addWithOverflow(
-                    decimal[offset],
-                    decimal[offset + 1],
-                    otherDecimal[otherOffset],
-                    otherDecimal[otherOffset + 1],
+                    decimal[decimalOffset],
+                    decimal[decimalOffset + 1],
+                    otherDecimal[otherDecimalOffset],
+                    otherDecimal[otherDecimalOffset + 1],
                     decimal,
-                    offset);
-            state.addOverflow(overflow + otherState.getOverflow());
+                    decimalOffset);
+            counterState.setValue(counterState.getValue() + otherCounterState.getValue());
+            if (overflow != 0 || !otherOverflowState.isNull()) {
```

Comment:
What about checking `overflow.isNull`? Do we miss that (missing test)?
You can do:
```
overflowState.setValue(overflowState.getValue() + overflow + otherOverflowState.getValue());
overflowState.setNull(overflowState != 0 | !otherOverflowState.isNull() | !overflow.isNull());
```

btw: use `|` instead of `||` for simple conditions to avoid branching